### ISSUE TITLE
BUG: Accelerate detection fails on at least one machine

### ISSUE THREAD

**Issue by mattip** [numpy/numpy#15946]
Copied from  [this comment](https://github.com/numpy/numpy/pull/15759#discussion_r406232652) in gh-15759:

> I don't get an error on import. The code added in` _mac_os_check()` in `__init__.py` in doesn't crash when I run it

However, the system in question (Mac OS 10.12.6 (Sierra)) does fail one of the tests:
```
>>> import numpy as np
>>> from numpy.lib.tests.test_polynomial import TestPolynomial
>>> tst = TestPolynomial()
>>> tst.test_polyfit()
python3(18846,0x7fffa26593c0) malloc: *** mach_vm_map(size=18446744073373515776) failed (error code=3)
*** error: can't allocate region
*** set a breakpoint in malloc_error_break to debug
init_dgelsd failed init
<stdin>:1: RankWarning: Polyfit may be poorly conditioned
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/warren/py381/lib/python3.8/site-packages/numpy-1.19.0.dev0+42228fc-py3.8-macosx-10.12-x86_64.egg/numpy/lib/tests/test_polynomial.py", line 171, in test_polyfit
    assert_allclose(mean.std(), 0.5, atol=0.01)
  File "/Users/warren/py381/lib/python3.8/site-packages/numpy-1.19.0.dev0+42228fc-py3.8-macosx-10.12-x86_64.egg/numpy/testing/_private/utils.py", line 1527, in assert_allclose
    assert_array_compare(compare, actual, desired, err_msg=str(err_msg),
  File "/Users/warren/py381/lib/python3.8/site-packages/numpy-1.19.0.dev0+42228fc-py3.8-macosx-10.12-x86_64.egg/numpy/testing/_private/utils.py", line 840, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Not equal to tolerance rtol=1e-07, atol=0.01

Mismatched elements: 1 / 1 (100%)
Max absolute difference: 0.5
Max relative difference: 1.
 x: array(0.)
 y: array(0.5)
>>> np.__version__  # Current master branch
'1.19.0.dev0+42228fc'
```

The test at import is meant to catch the warning emitted during that test. @WarrenWeckesser which of the allocations in that test triggers the warning? @vrakesh could you verify the warning is generated on your machine as well and adapt the test in `_mac_os_check` accordingly to catch more cases?

**Comment by WarrenWeckesser** [comment]
@mattip, just a "heads up" about that failing test:  I had to include `import numpy as np` first.  If I just ran
```
>>> from numpy.lib.tests.test_polynomial import TestPolynomial
>>> tst = TestPolynomial()
>>> tst.test_polyfit()
```
in a newly opened Python shell, the test did not crash.  So I don't know how reliable that test is.

(I'd also like to know if anyone else sees the behavior that I've seen.  I'd hate to be holding back progress on these Mac issues because of some strange quirk of my 7 year old laptop.)

**Comment by mattip** [comment]
@WarrenWeckesser if you use the code in gh-15759 (which removes a fix for Accelerate) does it change anything? Without that fix do all the tests pass on your machine? Perhaps your version of Accelerate has different bugs, we should have a test (or tests) for a range of Accelerate failures, and you could suggest an appropriate test.

**Comment by vrakesh** [comment]
@mattip, Let me dig deeper into this, will verify the above test case on my mac machine, will add more details here asap.

**Comment by vrakesh** [comment]
I tried to reproduce it on my Mac (Mojave) using the most recent master commit

I am still able to catch the issue

```python
from numpy.lib.tests.test_polynomial import TestPolynomial
python(62147,0x116cdd5c0) malloc: can't allocate region
*** mach_vm_map(size=18446744072152907776) failed (error code=3)
python(62147,0x116cdd5c0) malloc: *** set a breakpoint in malloc_error_break to debug
init_dgelsd failed init
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/rakvas/Workspace/numpy/numpy/__init__.py", line 286, in <module>
    raise RuntimeError(msg)
RuntimeError: Polyfit sanity test emitted a warning, most likely due to using a buggy Accelerate backend. If you compiled yourself, see site.cfg.example for information. Otherwise report this to the vendor that provided NumPy.
RankWarning: Polyfit may be poorly conditioned
>>> 
```

This is default compilation, so accelerate is used directly

After using remove-accelerate2 branch from [gh-15759](https://github.com/numpy/numpy/pull/15759)

The import still fails on my mac since the default BLAS on my mac , is a symlink to Accelereate

```python
git branch
  mac_os_test
  master
* remote-accelerate2
(numpyenv)$ python
Python 3.7.4 (default, Oct 12 2019, 18:55:28) 
[Clang 11.0.0 (clang-1100.0.33.8)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
python(68563,0x10c2fd5c0) malloc: can't allocate region
*** mach_vm_map(size=18446744072252174336) failed (error code=3)
python(68563,0x10c2fd5c0) malloc: *** set a breakpoint in malloc_error_break to debug
init_dgelsd failed init
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/vrakesh/Workspace/numpy/numpy/__init__.py", line 286, in <module>
    raise RuntimeError(msg)
RuntimeError: Polyfit sanity test emitted a warning, most likely due to using a buggy Accelerate backend. If you compiled yourself, see site.cfg.example for information. Otherwise report this to the vendor that provided NumPy.
RankWarning: Polyfit may be poorly conditioned

>>> 
```

```python
$ python
Python 3.7.4 (default, Oct 12 2019, 18:55:28) 
[Clang 11.0.0 (clang-1100.0.33.8)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from numpy.lib.tests.test_polynomial import TestPolynomial
python(70838,0x10d7595c0) malloc: can't allocate region
*** mach_vm_map(size=18446744072460599296) failed (error code=3)
python(70838,0x10d7595c0) malloc: *** set a breakpoint in malloc_error_break to debug
init_dgelsd failed init
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/vrakesh/Workspace/numpy/numpy/__init__.py", line 286, in <module>
    raise RuntimeError(msg)
RuntimeError: Polyfit sanity test emitted a warning, most likely due to using a buggy Accelerate backend. If you compiled yourself, see site.cfg.example for information. Otherwise report this to the vendor that provided NumPy.
RankWarning: Polyfit may be poorly conditioned

>>> 
```

``bash
ls -l /usr/lib/libblas.dylib 
lrwxr-xr-x  1 root  wheel  116 Sep 25  2019 /usr/lib/libblas.dylib -> ../../System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib

```
Which is expected behavior.


@WarrenWeckesser 

Wondering if default blas on 10.12 on Sierra is a symlink to accelerate, requesting information on the same.

This is very strange, indeed

**Comment by WarrenWeckesser** [comment]
For the record, here are details of my system:

* MacBook Pro late 2013, running MacOS 10.12.6 (Sierra)
* XCode info:

      $ xcodebuild -version
      Xcode 9.2
      Build version 9C40b
      $ xcode-select -version
      xcode-select version 2347.

* Python 3.8.1, built from source.

      $ python3
      Python 3.8.1 (default, Feb  8 2020, 12:50:00) 
      [Clang 9.0.0 (clang-900.0.39.2)] on darwin
      Type "help", "copyright", "credits" or "license" for more information.
      >>> 

With master (currently 1.19.0.dev0+b489287), the polyfit test might pass or it might crash--it depends.  Currently, if I run the full test suite from the command line (i.e. `python3 runtests.py -n -m full`), all the tests pass.  But if I run just the `lib` tests from within a Python shell...

    >>> np.lib.test('full')
    NumPy version 1.19.0.dev0+b489287
    NumPy relaxed strides checking option: True
    ........................................................................................................ [  5%]
    <... snip ...>
    ........................................................................................................ [ 84%]
    ...........................................F.................................................x.......... [ 89%]
    ........................................................................................................ [ 94%]
    ........................................................................................................ [ 99%]
    ...                                                                                                      [100%]
    =================================================== FAILURES ===================================================
    _________________________________________ TestPolynomial.test_polyfit __________________________________________

    self = <numpy.lib.tests.test_polynomial.TestPolynomial object at 0x11197a1f0>

    <... snip ...>

    py381/lib/python3.8/site-packages/numpy-1.19.0.dev0+b489287-py3.8-macosx-10.12-x86_64.egg/numpy/lib/tests/test_polynomial.py:181: AssertionError
    --------------------------------------------- Captured stderr call ---------------------------------------------
    python3(88704,0x7fffa26593c0) malloc: *** mach_vm_map(size=18446744065215049728) failed (error code=3)
    *** error: can't allocate region
    *** set a breakpoint in malloc_error_break to debug
    init_dgelsd failed init
    1 failed, 1973 passed, 2 skipped, 3 xfailed in 24.24s
    False
    >>> 

With master (currently 1.19.0.dev0+b489287):

    $ otool -L core/_multiarray_umath.cpython-38-darwin.so 
    core/_multiarray_umath.cpython-38-darwin.so:
        /System/Library/Frameworks/Accelerate.framework/Versions/A/Accelerate (compatibility version 1.0.0, current version 4.0.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1238.60.2)

    $ otool -L linalg/_umath_linalg.cpython-38-darwin.so 
    linalg/_umath_linalg.cpython-38-darwin.so:
        /System/Library/Frameworks/Accelerate.framework/Versions/A/Accelerate (compatibility version 1.0.0, current version 4.0.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1238.60.2)


With gh-15759 (1.19.0.dev0+9fdf2ec), the polyfit test crashes when the full test suite is run from the command line.

With gh-15759, it looks like the multiarray umath shared library no longer links to Accelerate, but the library in linalg still does:

    $ otool -L core/_multiarray_umath.cpython-38-darwin.so 
    core/_multiarray_umath.cpython-38-darwin.so:
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1238.60.2)

    $ otool -L linalg/_umath_linalg.cpython-38-darwin.so 
    linalg/_umath_linalg.cpython-38-darwin.so:
        /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libLAPACK.dylib (compatibility version 1.0.0, current version 1.0.0)
        /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib (compatibility version 1.0.0, current version 1.0.0)
        /usr/local/gcc-6.1.0/lib/libgfortran.3.dylib (compatibility version 4.0.0, current version 4.0.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1238.60.2)
        /usr/local/gcc-6.1.0/lib/libgcc_s.1.dylib (compatibility version 1.0.0, current version 1.0.0)
        /usr/local/gcc-6.1.0/lib/libquadmath.0.dylib (compatibility version 1.0.0, current version 1.0.0)

I don't know why those gcc-6.1.0 links are showing up there.  I have built gcc 6.1.0 from source, but that's not the C compiler that I'm using here.  The only command from that gcc build that is on my path is gfortran.

**Comment by mattip** [comment]
Maybe the difference between the two runs is the order of loading. If `_multiarray_umath` is loaded first, perhaps it prevents the Accelerate-tainted blas routines from `_umath_linalg` from loading. Why is `_umath_linalg` able to find Accelerate, I thought gh-15759 would have prevented that unless it is a symlink. Could you show the `build.log` lines where it finds Accelerate?

**Comment by WarrenWeckesser** [comment]
On my system (described above):

```
$ cd /usr/lib
/usr/lib$ ls -l libcblas.dylib libblas.dylib liblapack.dylib 
lrwxr-xr-x  1 root  wheel  116 Jul 17  2019 libblas.dylib@ -> ../../System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
lrwxr-xr-x  1 root  wheel  116 Jul 17  2019 libcblas.dylib@ -> ../../System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
lrwxr-xr-x  1 root  wheel  118 Jul 17  2019 liblapack.dylib@ -> ../../System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libLAPACK.dylib
```

**Comment by mattip** [comment]
So then why is it not used when building _multiarray_umath? Could you figure out why  `have_blas` in `numpy/core/setup.py` does not find a blas, but `use_lapack_lite` in `numpy/linalg/setup.py`does find a lapack?

**Comment by WarrenWeckesser** [comment]
@mattip: I put my build.log in a gist: https://gist.github.com/WarrenWeckesser/24373277b8562f922b4e72a5ed5d9a88

**Comment by WarrenWeckesser** [comment]
> Could you figure out why have_blas in numpy/core/setup.py does not find a blas, but use_lapack_lite in numpy/linalg/setup.pydoes find a lapack?

Those values are determined by the dictionaries returned by `get_info('blas_opt', 0)` and `get_info('lapack_opt', 0)`.  Here are those values:

```
>>> from numpy.distutils.system_info import get_info
>>> blas_info = get_info('blas_opt', 0)
/Users/warren/py381/lib/python3.8/site-packages/numpy-1.19.0.dev0+9fdf2ec-py3.8-macosx-10.12-x86_64.egg/numpy/distutils/system_info.py:1913: UserWarning: 
    Optimized (vendor) Blas libraries are not found.
    Falls back to netlib Blas library which has worse performance.
    A better performance should be easily gained by switching
    Blas library.
  if self._calc_info(blas):
/var/folders/p9/j_v_3wvd2bx5ftq544pvvf1w0000gn/T/tmpxj6061d3/source.c:1:10: fatal error: 'cblas.h' file not found
#include <cblas.h>
         ^~~~~~~~~
1 error generated.
>>> blas_info
{'define_macros': [('NO_ATLAS_INFO', 1)], 'libraries': ['blas', 'blas'], 'library_dirs': ['/usr/lib'], 'include_dirs': ['/usr/local/include', '/usr/include', '/Users/warren/py381/include']}
>>> lapack_info = get_info('lapack_opt', 0)
>>> lapack_info
{'libraries': ['lapack', 'lapack', 'blas', 'blas'], 'library_dirs': ['/usr/lib'], 'language': 'f77', 'define_macros': [('NO_ATLAS_INFO', 1)], 'include_dirs': ['/usr/local/include', '/usr/include', '/Users/warren/py381/include']}
```

`have_blas` is computed as

    have_blas = blas_info and ('HAVE_CBLAS', None) in blas_info.get('define_macros', [])

and `('HAVE_CBLAS', None)` is not in the `'define_macros'` list, so `have_blas` is False.

`use_lapack_lite` is simply

```
    use_lapack_lite = not lapack_info
```
and `lapack_info` is a non-empty dictionary, so `use_lapack_lite` is False.

I can dig deeper, but pointers to the most likely places to dig would help.

**Comment by mattip** [comment]
It seems we demand more from the `blas_info.calc_info` than from `lapack_info.calc_info`. `blas_info` must pass `get_cblas_libs` which tries to compile a snippet of code, where `lapack_info` just checks the existence of the libraries. So on your system, the blas_info compilation fails. On a more update macOS I think the blas test passes, and Accelerate is used for both.

Do we ever want to allow a situation where there is no BLAS but there is a LAPACK? What happens if you try this fix:
```
diff --git a/numpy/linalg/setup.py b/numpy/linalg/setup.py
index acfab0a68..435049cc4 100644
--- a/numpy/linalg/setup.py
+++ b/numpy/linalg/setup.py
@@ -26,10 +26,12 @@ def configuration(parent_package='', top_path=None):
 
     if os.environ.get('NPY_USE_BLAS_ILP64', "0") != "0":
         lapack_info = get_info('lapack_ilp64_opt', 2)
+        blas_info = get_info('blas_ilp64_opt', 2)
     else:
         lapack_info = get_info('lapack_opt', 0)  # and {}
+        blas_info = get_info('blas_opt', 0)
 
-    use_lapack_lite = not lapack_info
+    use_lapack_lite = not blas_info or not lapack_info
 
     if use_lapack_lite:
         # This makes numpy.distutils write the fact that lapack_lite
```

**Comment by WarrenWeckesser** [comment]
Both `blas_info` and `lapack_info` are non-empty dictionaries, so `not blas_info or not lapack_info` will be False.  Did you mean something like `not (have_blas and lapack_info)` (equivalently `not have_blas or not lapack_info`), where `have_blas` is computed as in `core/setup.py`?

**Comment by mattip** [comment]
> Did you mean ... ?

Yes

**Comment by WarrenWeckesser** [comment]
If I change the relevant code in linalg/setup.py to

```
    if os.environ.get('NPY_USE_BLAS_ILP64', "0") != "0":
        lapack_info = get_info('lapack_ilp64_opt', 2)
        blas_info = get_info('blas_ilp64_opt', 2)
    else:
        lapack_info = get_info('lapack_opt', 0)  # and {}
        blas_info = get_info('blas_opt', 0)

    have_blas = blas_info and (('HAVE_CBLAS', None) in
                               blas_info.get('define_macros', []))
    use_lapack_lite = not (have_blas and lapack_info)
```
then there are no links to Accelerate in the shared libs, and the polyfit test does not crash, either in the full test suite or when run individually.  :tada:

**Comment by pv** [comment]
It would be better to do the Accelerate check (i.e. don't consider Accelerate blas/lapack as generic blas) in numpy.distutils somehow, in order to help downstream projects (e.g. scipy) to not crash at runtime.

**Comment by pv** [comment]
I guess this could even be as simple as checking whether (i) platform is on darwin, and (ii) the Blas/lapack files are symlinks that point to a path that starts with `/System/Library/Frameworks/Accelerate.framework/`

**Comment by WarrenWeckesser** [comment]
> I guess this could even be as simple as checking whether (i) platform is on darwin, and (ii) the Blas/lapack files are symlinks that point to a path that starts with /System/Library/Frameworks/Accelerate.framework/

Sure, I'll give that a try.

**Comment by WarrenWeckesser** [comment]
>>    I guess this could even be as simple as checking whether (i) platform is on darwin, and (ii) the Blas/lapack files are symlinks that point to a path that starts with /System/Library/Frameworks/Accelerate.framework/

> Sure, I'll give that a try.

I implemented this in https://github.com/WarrenWeckesser/numpy/commit/b0c375ae5db1c0f02f99ce452db7922ed5685a2c, a commit in an extension of Matti's branch: https://github.com/WarrenWeckesser/numpy/tree/mattip-remove-accelerate2

**Comment by mattip** [comment]
Is there a reason it needs to be related to gh-15759 or could it be a stand-alone PR?

**Comment by WarrenWeckesser** [comment]
@mattip, I was thinking the same thing after I pushed the commit.  It looks like it can be stand-alone, so I'll create an independent PR.

**Comment by WarrenWeckesser** [comment]
Independent pull request at https://github.com/numpy/numpy/pull/15966

**Comment by rossbar** [comment]
Quick update: the changes originally in #15966 are now handled by #15759. Perhaps it's worth testing that this issue is resolved in 82398b3.

**Comment by mattip** [comment]
Closing
